### ISSUE TITLE
Union of tuples with literal types doesn't narrow as expected

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6691]
### Discussed in https://github.com/microsoft/pyright/discussions/6686

<div type='discussions-op-text'>

<sup>Originally posted by **bwburnsides** December  8, 2023</sup>
Hello!

I am working on a bootstrap compiler in Python 3.12 and relying on pyright for static type analysis. I've just ran into an issue where the inferred type of a match target is not what I expect it to be. Here's my setup:
```
class ResultKind(enum.Enum):
    Match = enum.auto()
    NoMatch = enum.auto()
    Error = enum.auto()

class ParseError:
    ...

type ParseResult[T] = (
    Lit[ResultKind.NoMatch]
    | tuple[Lit[ResultKind.Match], T]
    | tuple[Lit[ResultKind.Error], ParseError]
)

def test_parse() -> ParseResult[int]:
    ...

match test_parse():
    case ResultKind.NoMatch:
        ...
    case (ResultKind.Match, item):
        ...
    case (ResultKind.Error, err):
        ...
```
In the `(ResultKind.Match, item)` case, I expect `item` to be of type `int`. However, pyright finds it to be `int | ParseError`. And indeed if I swap the last two cases, it'll find `err` to be of type `int | ParseError`. Given the type alias explicitly mapping `ResultKind.Match` to `T`, and `ParseKind.Error` to `ParseError`, this is unexpected to me.

If I only include the `ResultKind.NoMatch` case, then pyright gives the warning I expect, which seems to contrast with the behavior show above.
```
Unhandled type: "tuple[Literal[ResultKind.Match], int] | tuple[Literal[ResultKind.Error], ParseError]"
```

What am I missing here?</div>

**Comment by erictraut** [comment]
I've enhanced the type narrowing logic for sequential patterns to support tuple narrowing based on subpattern matching. This is a pretty complex and sophisticated type narrowing case. There are still edge cases that pyright doesn't support here (e.g. if the tuple has an indeterminate length), but this covers the more common cases. Incidentally, mypy's type narrowing logic doesn't handle this case.

This will be addressed in the next release.

**Comment by bwburnsides** [comment]
Thanks for the quick fix!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.340, which I just published. It will be included in a future release of pylance.
